File: integration-tests/complexTypes/src/main/resources/dataProcessor.js
/**
 * JavaScript implementation of DataProcessor interface.
 * Uses context methods from DataProcessorContext for complex operations.
 */

function createPerson(name, age, street, city, zipCode) {
    // Use context method to create address
    const address = DataProcessor_Builtins.createAddress(street, city, zipCode);

    return {
        name: name,
        age: age,
        address: address
    };
}

function processPerson(person) {
    // Validate the person using context method
    if (!DataProcessor_Builtins.validatePerson(person)) {
        return {
            value: 0.0,
            message: "Invalid person data",
            timestamp: null,
            tags: ["error", "validation"]
        };
    }

    // Format address using context method
    const formattedAddress = DataProcessor_Builtins.formatAddress(person);

    return {
        value: person.age * 1.5,
        message: "Processed: " + person.name + " at " + formattedAddress,
        timestamp: null,
        tags: ["success", "processed"]
    };
}

export {
    processPerson,
    createPerson
};

File: integration-tests/complexTypes/src/main/java/com/example/DataProcessor.java
package com.example;

import io.roastedroot.quickjs4j.annotations.ScriptInterface;

/**
 * Script interface for processing data with complex types.
 * JavaScript implementations will receive TypeScript type definitions for context methods.
 */
@ScriptInterface(context = DataProcessorContext.class)
public interface DataProcessor {

    /**
     * Processes a person and returns a result.
     *
     * @param person the person to process
     * @return calculation result
     */
    CalculationResult processPerson(Person person);

    /**
     * Creates a person from basic information.
     *
     * @param name the person's name
     * @param age the person's age
     * @param street the street address
     * @param city the city
     * @param zipCode the zip code
     * @return a new Person object
     */
    Person createPerson(String name, int age, String street, String city, String zipCode);
}
